chasm/tree.go

```diff
@@ -668,10 +687,126 @@ func (n *Node) AddTask(
 }
 
+func (n *Node) closeTransactionGenratePhysicalSideEffectTasks() error {
+	entityKey := n.backend.GetWorkflowKey()
+
+	for encodedPath, updatedNode := range n.mutation.UpdatedNodes {
+		componentAttr := updatedNode.GetMetadata().GetComponentAttributes()
+		if componentAttr == nil {
+			continue
+		}
+
+		sideEffectTasks := componentAttr.GetSideEffectTasks()
+		for idx := len(sideEffectTasks) - 1; idx >= 0; idx-- {
+			sideEffectTask := sideEffectTasks[idx]
+			if sideEffectTask.PhysicalTaskStatus == physicalTaskStatusCreated {
+				break
+			}
+
+			category, err := taskCategory(sideEffectTask)
+			if err != nil {
+				return err
+			}
+
+			physicalTask := &tasks.ChasmTask{
+				WorkflowKey:         entityKey,
+				VisibilityTimestamp: sideEffectTask.ScheduledTime.AsTime(),
+				Destination:         sideEffectTask.Destination,
+				Category:            category,
+				Info: &persistencespb.ChasmTaskInfo{
+					Ref: &persistencespb.ChasmComponentRef{
+						ComponentInitialVersionedTransition:    updatedNode.Metadata.InitialVersionedTransition,
+						ComponentLastUpdateVersionedTransition: updatedNode.Metadata.LastUpdateVersionedTransition,
+						Path:                                   encodedPath,
+					},
+					Type: sideEffectTask.Type,
+					Data: sideEffectTask.Data,
+				},
+			}
+			n.backend.AddTasks(physicalTask)
+			sideEffectTask.PhysicalTaskStatus = physicalTaskStatusCreated
+		}
+	}
+
+	return nil
+}
+
+func (n *Node) closeTransactionGeneratePhysicalPureTask() error {
+	var firstPureTask *persistencespb.ChasmComponentAttributes_Task
+	var firstTaskNode *Node
+	if err := n.walk(func(node *Node) error {
+		componentAttr := node.serializedNode.GetMetadata().GetComponentAttributes()
+		if componentAttr == nil {
+			return nil
+		}
+
+		pureTasks := componentAttr.GetPureTasks()
+		if len(pureTasks) == 0 {
+			return nil
+		}
+
+		if firstPureTask == nil ||
+			comparePureTasks(pureTasks[0], firstPureTask) < 0 {
+			firstPureTask = pureTasks[0]
+			firstTaskNode = node
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if firstPureTask == nil || firstPureTask.PhysicalTaskStatus == physicalTaskStatusCreated {
+		return nil
+	}
+
+	n.backend.AddTasks(&tasks.ChasmTaskPure{
+		WorkflowKey:         n.backend.GetWorkflowKey(),
+		VisibilityTimestamp: firstPureTask.ScheduledTime.AsTime(),
+		Category:            tasks.CategoryTimer,
+	})
+
+	// We need to persist the task status change as well, so add the node
+	// to the list of updated nodes.
+	// However, since task status is a cluster local field, we don't really
+	// update LastUpdateVersionedTransition for this node, and the change won't be replicated.
+	firstPureTask.PhysicalTaskStatus = physicalTaskStatusCreated
+	encodedPath, err := firstTaskNode.encodedPath()
+	if err != nil {
+		return err
+	}
+	n.mutation.UpdatedNodes[encodedPath] = firstTaskNode.serializedNode
+	return nil
+}
+
+func (n *Node) walk(
+	visitor func(node *Node) error,
+) error {
```

Comment:
I think this can be an iterator too.

Reply:
I guess the walk() method itself is already kinda an iterator :)